borsh/src/de/borrowed.rs

```diff
@@ -0,0 +1,85 @@
+use crate::{
+    maybestd::{
+        io::{Error, ErrorKind, Result},
+        vec::Vec,
+    },
+    BorshDeserialize,
+};
+
+use super::{ERROR_NOT_ALL_BYTES_READ, ERROR_UNEXPECTED_LENGTH_OF_INPUT};
+
+/// A borrowed data-structure that can be de-serialized from binary format by NBOR.
+pub trait BorshBorrowedDeserialize<'de>: Sized {
+    /// Deserializes this instance from a given slice of bytes.
+    /// Updates the buffer to point at the remaining bytes.
+    fn deserialize(buf: &mut &'de [u8]) -> Result<Self>;
+
+    /// Deserialize this instance from a slice of bytes.
+    fn try_from_slice(v: &'de [u8]) -> Result<Self> {
+        let mut v_mut = v;
+        let result = Self::deserialize(&mut v_mut)?;
+        if !v_mut.is_empty() {
+            return Err(Error::new(ErrorKind::InvalidData, ERROR_NOT_ALL_BYTES_READ));
+        }
+        Ok(result)
+    }
+
+    #[inline]
+    #[doc(hidden)]
+    fn vec_from_bytes(len: u32, buf: &mut &'de [u8]) -> Result<Option<Vec<Self>>> {
+        let _ = len;
+        let _ = buf;
+        Ok(None)
+    }
+
+    #[inline]
+    #[doc(hidden)]
+    fn array_from_bytes<const N: usize>(buf: &mut &'de [u8]) -> Result<Option<[Self; N]>> {
+        let _ = buf;
+        Ok(None)
+    }
+}
+
+impl<'de, T: BorshDeserialize> BorshBorrowedDeserialize<'de> for T {
+    fn deserialize(buf: &mut &'de [u8]) -> Result<Self> {
+        T::deserialize(buf)
+    }
```

Comment:
I think you also need to redirect `vec_from_bytes` and `array_from_bytes` to enable the optimizations Borsh uses them for.

Reply:
Otherwise, we have no use for them, remove them from the new trait.